src/backend/kvm/mod.rs

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::Loader;
+use crate::backend::kvm::data::{dev_kvm, kvm_version};
+use crate::backend::kvm::mem::Region;
+use anyhow::Result;
+pub use kvm_bindings::kvm_userspace_memory_region as KvmUserspaceMemoryRegion;
+use kvm_ioctls::Kvm;
+use kvm_ioctls::{VcpuFd, VmFd};
+use lset::Span;
+use mmarinus::{perms, Kind, Map};
+use primordial::Page;
+use std::collections::VecDeque;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use x86_64::VirtAddr;
+
+mod builder;
+mod config;
+mod data;
+mod hasher;
+mod mem;
+mod thread;
+
+struct Keep {
+    kvm_fd: Kvm,
+    vm_fd: VmFd,
+    cpu_fds: VecDeque<VcpuFd>,
+    sallyport_blocks: Span<VirtAddr, NonZeroUsize>,
+    regions: Vec<Region>,
+}
+
+impl Keep {
+    pub fn add_memory(&mut self, pages: usize) -> Result<i64> {
+        let mem_size = pages * Page::SIZE;
+        let last_region = self.regions.last().unwrap().as_guest();
+
+        let map = Map::map(mem_size as usize)
+            .anywhere()
+            .anonymously()
+            .known::<perms::ReadWrite>(Kind::Private)?;
+
+        let region_start = map.addr();
+        let region = KvmUserspaceMemoryRegion {
+            slot: self.regions.len() as _,
+            flags: 0,
+            guest_phys_addr: last_region.start.as_u64() + last_region.count,
+            memory_size: mem_size as _,
+            userspace_addr: region_start as _,
+        };
+
+        unsafe {
+            self.vm_fd.set_user_memory_region(region)?;
+        }
+
+        self.regions.push(Region::new(region, map));
+
+        Ok(region_start as _)
+    }
```

Comment:
This approach has several drawbacks that could be hard in the future:
1. The page size is fixed. This API has no way to specify huge pages.
2. All memory expansion MUST be contiguous.
3. Presuming that the region to be expanded is the last region imposes layout constraints.

I proposed a more flexible version of this in my version of the KVM backend. Basically, instead of the guest having only one output parameter (`pages`) it should have:
* `page_size`: a hint at the preferred page size
* `pages`: the amount of memory to allocate
* `address`: the location to map the memory to (must be aligned to `page_size`)

I think the most important aspect of this interface is that the loader imposes no layout constraints on the shim; so we have complete flexibility in the future.

Reply:
:+1: agreed